src/Mod/Sketcher/App/planegcs/Constraints.cpp

```diff
@@ -462,6 +462,123 @@ double ConstraintSlopeAtBSplineKnot::grad(double *param)
     return scale * result;
 }
 
+// Point On BSpline
+
+ConstraintPointOnBSpline::ConstraintPointOnBSpline(double* point, double* initparam, int coordidx, BSpline& b)
+    : bsp(b)
+{
+    // This is always going to be true
+    numpoints = bsp.degree + 1;
+
+    pvec.reserve(2 + 2*b.poles.size());
+    pvec.push_back(point);
+    pvec.push_back(initparam);
+
+    setStartPole(*initparam);
+
+    for (size_t i = 0; i < b.poles.size(); ++i) {
+        if (coordidx == 0)
+            pvec.push_back(b.poles[i].x);
+        else
+            pvec.push_back(b.poles[i].y);
+    }
+    for (size_t i = 0; i < b.weights.size(); ++i)
+        pvec.push_back(b.weights[i]);
+
+    if (bsp.flattenedknots.empty())
+        bsp.setupFlattenedKnots();
+
+    origpvec = pvec;
+    rescale();
+}
+
+ConstraintType ConstraintPointOnBSpline::getTypeId()
+{
+    return PointOnBSpline;
+}
+
+void ConstraintPointOnBSpline::setStartPole(double u)
+{
+    // The startpole logic is repeated in a lot of places,
+    // for example in GCS and slope at knot
+    // find relevant poles
+    startpole = 0;
+    for (size_t j = 1; j < bsp.mult.size() && *(bsp.knots[j]) <= u; ++j)
+        startpole += bsp.mult[j];
+    if (!bsp.periodic && startpole >= bsp.poles.size())
+        startpole = bsp.poles.size() - bsp.degree - 1;
+}
+
+void ConstraintPointOnBSpline::rescale(double coef)
+{
+    scale = coef * 1.0;
+}
+
+double ConstraintPointOnBSpline::error()
+{
+    if (*theparam() < bsp.flattenedknots[startpole + bsp.degree] ||
+        *theparam() > bsp.flattenedknots[startpole + bsp.degree + 1])
+        setStartPole(*theparam());
+
+    double sum = 0;
+    double wsum = 0;
+
+    // TODO: maybe make it global so it doesn't have to be created every time
+    VEC_D d(numpoints);
```

Comment:
@AjinkyaDahale 

The penalty is probably not much. If "numpoints" is always the same value, I wonder if making it static would be an option. It would have local static storage to the function and would be release upon destruction of the object.

I will be merging it like this now,  but think about it.

Reply:
That's a good idea. TODO for the morning.